File: sdk-go/verify_test.go
package ztxbas

import (
	"context"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync/atomic"
	"testing"
	"time"
)

// testSigner is a tiny in-test ES256 signer that mirrors the server's
// signing scheme. Kept here (not in the SDK) so the SDK itself has no
// signing surface — RPs only ever verify.
type testSigner struct {
	priv *ecdsa.PrivateKey
	kid  string
}

func newTestSigner(t *testing.T) *testSigner {
	t.Helper()
	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	return &testSigner{priv: priv, kid: "test-kid"}
}

func b64u(b []byte) string { return base64.RawURLEncoding.EncodeToString(b) }

func (s *testSigner) sign(claims Claims) string {
	hdr, _ := json.Marshal(map[string]string{"alg": "ES256", "typ": "JWT", "kid": s.kid})
	pl, _ := json.Marshal(claims)
	input := b64u(hdr) + "." + b64u(pl)
	sum := sha256.Sum256([]byte(input))
	r, ss, _ := ecdsa.Sign(rand.Reader, s.priv, sum[:])
	sig := make([]byte, 64)
	rb, sb := r.Bytes(), ss.Bytes()
	copy(sig[32-len(rb):32], rb)
	copy(sig[64-len(sb):64], sb)
	return input + "." + b64u(sig)
}

func (s *testSigner) jwks() []byte {
	pub := s.priv.PublicKey
	byteLen := (pub.Curve.Params().BitSize + 7) / 8
	x, y := make([]byte, byteLen), make([]byte, byteLen)
	xb, yb := pub.X.Bytes(), pub.Y.Bytes()
	copy(x[byteLen-len(xb):], xb)
	copy(y[byteLen-len(yb):], yb)
	doc, _ := json.Marshal(map[string]any{
		"keys": []map[string]string{
			{"kty": "EC", "crv": "P-256", "alg": "ES256", "use": "sig", "kid": s.kid, "x": b64u(x), "y": b64u(y)},
		},
	})
	return doc
}

func jwksHandler(s *testSigner, hits *int64) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if hits != nil {
			atomic.AddInt64(hits, 1)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write(s.jwks())
	})
}

func TestVerifier_Happy(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()

	v, err := NewVerifier(ts.URL, WithExpectedIssuer("https://ztxbas.example.com"))
	if err != nil {
		t.Fatalf("NewVerifier: %v", err)
	}
	token := sig.sign(Claims{
		Issuer: "https://ztxbas.example.com", Subject: "alice@example.com",
		Audience: "https://app.example.com", Email: "alice@example.com",
		Origin: "https://app.example.com", ChallengeID: "c_1",
		IssuedAt: time.Now().Unix(), ExpiresAt: time.Now().Add(5 * time.Minute).Unix(),
	})
	claims, err := v.Verify(context.Background(), token)
	if err != nil {
		t.Fatalf("Verify: %v", err)
	}
	if claims.Email != "alice@example.com" || claims.Origin != "https://app.example.com" {
		t.Errorf("claims: %+v", claims)
	}
}

func TestVerifier_RejectsAlgNone(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL)

	// Handcraft an `alg: none` token with the same payload — the
	// classic algorithm-substitution attack. Must be rejected.
	hdr, _ := json.Marshal(map[string]string{"alg": "none", "typ": "JWT", "kid": sig.kid})
	pl, _ := json.Marshal(Claims{Email: "attacker@example.com", ExpiresAt: time.Now().Add(time.Hour).Unix()})
	token := b64u(hdr) + "." + b64u(pl) + "."
	if _, err := v.Verify(context.Background(), token); err == nil {
		t.Fatal("expected error on alg=none token")
	} else if !strings.Contains(err.Error(), "alg") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestVerifier_RejectsExpired(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL)

	token := sig.sign(Claims{
		ExpiresAt: time.Now().Add(-1 * time.Hour).Unix(),
	})
	_, err := v.Verify(context.Background(), token)
	if err == nil || !strings.Contains(err.Error(), "expired") {
		t.Errorf("want expiry error, got %v", err)
	}
}

func TestVerifier_RejectsWrongIssuer(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL, WithExpectedIssuer("https://legit.example.com"))

	token := sig.sign(Claims{
		Issuer: "https://evil.example.com", ExpiresAt: time.Now().Add(time.Hour).Unix(),
	})
	_, err := v.Verify(context.Background(), token)
	if err == nil || !strings.Contains(err.Error(), "iss") {
		t.Errorf("want iss mismatch error, got %v", err)
	}
}

func TestVerifier_RejectsTamperedPayload(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL)

	orig := sig.sign(Claims{
		Email: "alice@example.com", ExpiresAt: time.Now().Add(time.Hour).Unix(),
	})
	parts := strings.Split(orig, ".")
	// Swap payload for a claim declaring a different email; signature
	// no longer verifies over the new signing input.
	forged, _ := json.Marshal(Claims{
		Email: "attacker@example.com", ExpiresAt: time.Now().Add(time.Hour).Unix(),
	})
	tampered := parts[0] + "." + b64u(forged) + "." + parts[2]
	if _, err := v.Verify(context.Background(), tampered); err == nil {
		t.Fatal("expected signature error on tampered payload")
	}
}

func TestVerifier_RejectsUnknownKID(t *testing.T) {
	sig := newTestSigner(t)
	sig.kid = "known-kid"
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL)

	// Sign with an entirely different key but advertise our known kid?
	// Simpler: sign with a rogue signer that declares a kid the JWKS
	// doesn't publish. Verifier should refuse.
	rogue := newTestSigner(t)
	rogue.kid = "unknown-kid"
	tok := rogue.sign(Claims{ExpiresAt: time.Now().Add(time.Hour).Unix()})
	if _, err := v.Verify(context.Background(), tok); err == nil {
		t.Fatal("expected error on unknown kid")
	}
}

func TestVerifier_RejectsOffCurvePublicKey(t *testing.T) {
	// Serve a JWKS whose (x, y) is not on P-256 — this catches an
	// invalid-curve attack where the verifier is tricked into doing
	// scalar math on a weak curve.
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		bad, _ := json.Marshal(map[string]any{
			"keys": []map[string]string{{
				"kty": "EC", "crv": "P-256", "alg": "ES256", "use": "sig", "kid": "bad",
				"x": b64u(big.NewInt(1).Bytes()), "y": b64u(big.NewInt(1).Bytes()),
			}},
		})
		_, _ = w.Write(bad)
	}))
	defer ts.Close()

	v, _ := NewVerifier(ts.URL)
	// Any well-formed ES256 token will do — verify should fail during
	// JWKS decode, not signature check.
	sig := newTestSigner(t)
	tok := sig.sign(Claims{ExpiresAt: time.Now().Add(time.Hour).Unix()})
	if _, err := v.Verify(context.Background(), tok); err == nil {
		t.Fatal("expected JWKS decode error for off-curve key")
	}
}

func TestVerifier_JWKSCacheReusesWithinTTL(t *testing.T) {
	sig := newTestSigner(t)
	var hits int64
	ts := httptest.NewServer(jwksHandler(sig, &hits))
	defer ts.Close()

	v, _ := NewVerifier(ts.URL)
	tok := sig.sign(Claims{ExpiresAt: time.Now().Add(time.Hour).Unix()})
	for i := 0; i < 3; i++ {
		if _, err := v.Verify(context.Background(), tok); err != nil {
			t.Fatalf("verify %d: %v", i, err)
		}
	}
	// First Verify fills the cache; subsequent ones must not refetch.
	if got := atomic.LoadInt64(&hits); got != 1 {
		t.Errorf("JWKS fetched %d times, want 1", got)
	}
}

func TestVerifier_MalformedToken(t *testing.T) {
	sig := newTestSigner(t)
	ts := httptest.NewServer(jwksHandler(sig, nil))
	defer ts.Close()
	v, _ := NewVerifier(ts.URL)

	cases := []string{"", "a.b", "a.b.c.d", "not-base64-!!!.also.bad"}
	for _, c := range cases {
		if _, err := v.Verify(context.Background(), c); err == nil {
			t.Errorf("expected error for %q", c)
		}
	}
}

// TestVerifier_ClientLazyBuild covers Client.VerifyJWT wiring — it
// should build a Verifier on first use pointing at the client's
// baseURL + /.well-known/jwks.json.
func TestVerifier_ClientLazyBuild(t *testing.T) {
	sig := newTestSigner(t)
	mux := http.NewServeMux()
	mux.Handle("/.well-known/jwks.json", jwksHandler(sig, nil))
	ts := httptest.NewServer(mux)
	defer ts.Close()

	c, err := New(ts.URL, "app_test", "sec")
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	tok := sig.sign(Claims{Email: "alice@example.com", ExpiresAt: time.Now().Add(time.Hour).Unix()})
	claims, err := c.VerifyJWT(context.Background(), tok)
	if err != nil {
		t.Fatalf("VerifyJWT: %v", err)
	}
	if claims.Email != "alice@example.com" {
		t.Errorf("claims: %+v", claims)
	}
}

// sanity: keep an import used only inside test bodies compiling.
var _ = fmt.Sprintf

File: sdk-node/test/verify.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
  createHash,
  createPrivateKey,
  createSign,
  generateKeyPairSync,
  KeyObject,
} from 'node:crypto';
import { createServer, Server } from 'node:http';
import { Verifier, JwtVerifyError, Client } from '../src/index.js';

/**
 * Tiny in-test ES256 signer that mirrors the server's signing scheme.
 * Lives in the test file (not the SDK) so the SDK itself has no
 * signing surface — RPs only ever verify.
 */
class TestSigner {
  readonly kid: string;
  readonly privateKey: KeyObject;
  readonly publicKeyJwk: { kty: string; crv: string; x: string; y: string };

  constructor(kid = 'test-kid') {
    this.kid = kid;
    const { privateKey, publicKey } = generateKeyPairSync('ec', { namedCurve: 'P-256' });
    this.privateKey = privateKey;
    const jwk = publicKey.export({ format: 'jwk' }) as { kty: string; crv: string; x: string; y: string };
    this.publicKeyJwk = jwk;
  }

  sign(claims: Record<string, unknown>): string {
    const header = { alg: 'ES256', typ: 'JWT', kid: this.kid };
    const enc = (o: unknown) => Buffer.from(JSON.stringify(o)).toString('base64url');
    const signingInput = `${enc(header)}.${enc(claims)}`;
    // Node's createSign returns DER; convert to raw r||s (JWS-compact form).
    const der = createSign('SHA256').update(signingInput).sign({
      key: this.privateKey,
      dsaEncoding: 'ieee-p1363', // raw r||s
    });
    return `${signingInput}.${Buffer.from(der).toString('base64url')}`;
  }

  jwksBody(): string {
    return JSON.stringify({
      keys: [{
        kty: 'EC', crv: 'P-256', alg: 'ES256', use: 'sig', kid: this.kid,
        x: this.publicKeyJwk.x, y: this.publicKeyJwk.y,
      }],
    });
  }
}

async function serveJwks(signer: TestSigner, onHit?: () => void): Promise<{ url: string; server: Server }> {
  const server = createServer((_, res) => {
    onHit?.();
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(signer.jwksBody());
  });
  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  const addr = server.address();
  const port = typeof addr === 'object' && addr ? addr.port : 0;
  return { url: `http://127.0.0.1:${port}/`, server };
}

test('verifier accepts a valid ES256 JWT and returns claims', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url, { expectedIssuer: 'https://ztxbas.example.com' });
    const now = Math.floor(Date.now() / 1000);
    const tok = sig.sign({
      iss: 'https://ztxbas.example.com',
      sub: 'alice@example.com',
      aud: 'https://app.example.com',
      email: 'alice@example.com',
      origin: 'https://app.example.com',
      challenge_id: 'c_1',
      iat: now,
      exp: now + 300,
    });
    const claims = await v.verify(tok);
    assert.equal(claims.email, 'alice@example.com');
    assert.equal(claims.origin, 'https://app.example.com');
  } finally {
    server.close();
  }
});

test('verifier rejects alg=none', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url);
    const enc = (o: unknown) => Buffer.from(JSON.stringify(o)).toString('base64url');
    const header = { alg: 'none', typ: 'JWT', kid: sig.kid };
    const payload = { email: 'attacker@example.com', exp: Math.floor(Date.now() / 1000) + 3600 };
    const tok = `${enc(header)}.${enc(payload)}.`;
    await assert.rejects(v.verify(tok), (e: Error) => e instanceof JwtVerifyError && /alg/i.test(e.message));
  } finally {
    server.close();
  }
});

test('verifier rejects expired token', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url);
    const tok = sig.sign({ exp: Math.floor(Date.now() / 1000) - 3600 });
    await assert.rejects(v.verify(tok), (e: Error) => /expired/i.test(e.message));
  } finally {
    server.close();
  }
});

test('verifier rejects wrong issuer', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url, { expectedIssuer: 'https://legit.example.com' });
    const tok = sig.sign({
      iss: 'https://evil.example.com',
      exp: Math.floor(Date.now() / 1000) + 3600,
    });
    await assert.rejects(v.verify(tok), (e: Error) => /iss/.test(e.message));
  } finally {
    server.close();
  }
});

test('verifier rejects tampered payload (signature invalid)', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url);
    const tok = sig.sign({ email: 'alice@example.com', exp: Math.floor(Date.now() / 1000) + 3600 });
    const [h, , s] = tok.split('.');
    const forged = Buffer.from(JSON.stringify({
      email: 'attacker@example.com', exp: Math.floor(Date.now() / 1000) + 3600,
    })).toString('base64url');
    await assert.rejects(v.verify(`${h}.${forged}.${s}`), (e: Error) => /signature invalid/i.test(e.message));
  } finally {
    server.close();
  }
});

test('verifier rejects unknown kid', async () => {
  const sig = new TestSigner('known-kid');
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url);
    const rogue = new TestSigner('unknown-kid');
    const tok = rogue.sign({ exp: Math.floor(Date.now() / 1000) + 3600 });
    await assert.rejects(v.verify(tok), (e: Error) => /kid/.test(e.message));
  } finally {
    server.close();
  }
});

test('verifier caches JWKS within TTL', async () => {
  let hits = 0;
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig, () => { hits += 1; });
  try {
    const v = new Verifier(url);
    const tok = sig.sign({ exp: Math.floor(Date.now() / 1000) + 3600 });
    for (let i = 0; i < 3; i++) {
      await v.verify(tok);
    }
    assert.equal(hits, 1, `expected 1 JWKS fetch, saw ${hits}`);
  } finally {
    server.close();
  }
});

test('verifier rejects malformed tokens', async () => {
  const sig = new TestSigner();
  const { url, server } = await serveJwks(sig);
  try {
    const v = new Verifier(url);
    for (const bad of ['', 'a.b', 'a.b.c.d', 'not-base64.also.bad']) {
      await assert.rejects(v.verify(bad));
    }
  } finally {
    server.close();
  }
});

test('Client.verifyJwt builds a Verifier at baseUrl/.well-known/jwks.json', async () => {
  const sig = new TestSigner();
  const server = createServer((req, res) => {
    if (req.url === '/.well-known/jwks.json') {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(sig.jwksBody());
      return;
    }
    res.writeHead(404);
    res.end();
  });
  await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
  const addr = server.address();
  const port = typeof addr === 'object' && addr ? addr.port : 0;
  try {
    const c = new Client(`http://127.0.0.1:${port}`, 'app_test', 'sec');
    const tok = sig.sign({ email: 'alice@example.com', exp: Math.floor(Date.now() / 1000) + 3600 });
    const claims = await c.verifyJwt(tok);
    assert.equal(claims.email, 'alice@example.com');
  } finally {
    server.close();
  }
});

// Suppress unused-import warning for createHash (kept available for future
// tests that want to compute their own JWKS document).
void createHash;
void createPrivateKey;
